src/core/reduce.rkt

```diff
@@ -25,13 +25,11 @@
   (set! fn-inverses
         (remove-duplicates
          (map
-          (λ (r) (resugar-program (rule-input r) (rule-otype r) #:full #f))
+          (λ (r) (rule-input r))
```

Comment:
This is great. IIRC this should also enable us to remove the `#:full` flag to `resugar-program`? I'm not 100% sure but worth checking. Perhaps we can also drop some `require`s in this file?